src/test/scala/mesosphere/marathon/core/task/tracker/impl/InstancesLoaderImplTest.scala

```diff
@@ -1,64 +1,153 @@
+        state.Instance.fromCoreInstance(app2Instance1)
+      )
+
+      f.instanceRepository.ids() returns Source(instances.map(_.instanceId)(collection.breakOut))
+      for (instance <- instances) {
+        f.instanceRepository.get(instance.instanceId) returns Future.successful(Some(instance))
+      }
 
-      And("our data is empty")
-      loaded.futureValue.allInstances should be(empty)
+      When("load is called")
+      val loaded = f.loader.load()
 
-      And("there are no more interactions")
-      f.verifyNoMoreInteractions()
+      Then("the resulting data includes an instance for app 2 with the latest run spec")
+      // we do not need to verify the mocked calls because the only way to get the data is to perform the calls
+      val expectedData = InstanceTracker.InstancesBySpec.forInstances(app1Instance1, app1Instance2, app2Instance1.copy(runSpec = app2Newer))
+      loaded.futureValue should equal(expectedData)
     }
 
-    "loading multiple tasks for multiple apps" in {
+    "ignore instances without a run spec" in {
       val f = new Fixture
 
       Given("instances for multiple runSpecs")
       val app1Id = PathId("/app1")
       val app1Instance1 = TestInstanceBuilder.newBuilder(app1Id).getInstance()
       val app1Instance2 = TestInstanceBuilder.newBuilder(app1Id).getInstance()
+      val app1 = app1Instance1.runSpec
+      f.groupRepository.runSpecVersion(eq(app1Id), eq(app1.version.toOffsetDateTime))(any) returns Future.successful(Some(app1))
+
       val app2Id = PathId("/app2")
       val app2Instance1 = TestInstanceBuilder.newBuilder(app2Id).getInstance()
-      val instances = Seq(app1Instance1, app1Instance2, app2Instance1)
+      f.groupRepository.runSpecVersion(eq(app2Id), eq(app2Instance1.runSpecVersion.toOffsetDateTime))(any) returns Future.successful(None)
+      f.groupRepository.latestRunSpec(eq(app2Id))(any, any) returns Future.successful(None)
+
+      val instances = Seq(
+        state.Instance.fromCoreInstance(app1Instance1),
+        state.Instance.fromCoreInstance(app1Instance2),
+        state.Instance.fromCoreInstance(app2Instance1)
+      )
 
       f.instanceRepository.ids() returns Source(instances.map(_.instanceId)(collection.breakOut))
       for (instance <- instances) {
         f.instanceRepository.get(instance.instanceId) returns Future.successful(Some(instance))
       }
 
+      f.instanceRepository.delete(app2Instance1.instanceId) returns Future.successful(Done)
+
       When("load is called")
       val loaded = f.loader.load()
 
-      Then("the resulting data is correct")
+      Then("the resulting data ")
       // we do not need to verify the mocked calls because the only way to get the data is to perform the calls
-      val expectedData = InstanceTracker.InstancesBySpec.forInstances(app1Instance1, app1Instance2, app2Instance1)
+      val expectedData = InstanceTracker.InstancesBySpec.forInstances(app1Instance1, app1Instance2)
```

Comment:
I think in this particular test we should verify that `expungeOrphanedInstance` was called, since that's the contract of current implementation